core/src/Streamly/Internal/Data/Unboxed.hs

```diff
@@ -257,132 +259,101 @@ writeWord8ArrayAsDouble# arr# i# =
 --
 -- This type class is similar to Storable. While Storable allows writing to a
 -- Ptr, this type class allows writing to a MutableByteArray#.
-class Unboxed a where
-    sizeOf :: a -> Int
-    alignment :: a -> Int
+type Unboxed a = (Unbox a, Storable a)
 
+class Storable a => Unbox a where
     -- Read an element of type "a" from a MutableByteArray given the byte
     -- index.
-    readByteArray :: ArrayContents a -> Int -> IO a
-    writeByteArray :: ArrayContents a -> Int -> a -> IO ()
-
-#define DERIVE_UNBOXED(_type, _constructor, _sizeOf, _alignment, _readArray, _writeArray) \
-instance Unboxed _type where {                                       \
-; {-# INLINE sizeOf #-}                                              \
-; sizeOf _ = _sizeOf                                                 \
-; {-# INLINE alignment #-}                                           \
-; alignment _ =  _alignment                                          \
-; {-# INLINE readByteArray #-}                                       \
-; readByteArray (ArrayContents mbarr) (I# n) = IO $ \s ->            \
+    box :: MutableByteArray a -> Int -> IO a
+    unbox :: MutableByteArray a -> Int -> a -> IO ()
```

Comment:
Add haddock, so that people can understand it when writing instances.

1. You have pushed the type class documentation above the type alias. Move it so that it belongs to the type class.
2. Add doc for "unbox" operation - "Write an element of type "a" to a MutableByteArray.